akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala

```diff
@@ -0,0 +1,394 @@
+package akka.typed.scaladsl.persistence
+
+import akka.typed
+
+import scala.concurrent.duration._
+import akka.typed.{ ActorRef, Behavior, ExtensibleBehavior, Signal, Terminated }
+import akka.typed.scaladsl.{ ActorContext, TimerScheduler }
+
+import scala.concurrent.ExecutionContext
+import scala.reflect.ClassTag
+
+class ApiTest {
+  object TypedPersistentActor {
+
+    sealed trait PersistentEffect[+Event]
+
+    case object PersistNothing extends PersistentEffect[Nothing]
+
+    case class Persist[Event](event: Event, callbacks: List[Event ⇒ Unit] = Nil) extends PersistentEffect[Event] {
+      def andThen(callback: Event ⇒ Unit) = Persist(event, callback :: callbacks)
+    }
+
+    case class Unhandled[Event]() extends PersistentEffect[Event]
+
+    class ActionHandler[Command: ClassTag, Event, State](val handler: ((Any, State, ActorContext[Command]) ⇒ PersistentEffect[Event])) {
+      def onSignal(signalHandler: (Any, State, ActorContext[Command]) ⇒ PersistentEffect[Event]): ActionHandler[Command, Event, State] =
+        new ActionHandler({
+          case (command: Command, state, ctx) ⇒ handler(command, state, ctx)
+          case (signal: Signal, state, ctx)   ⇒ signalHandler(signal, state, ctx)
+          case _                              ⇒ Unhandled[Event]()
+        })
+    }
+    object ActionHandler {
+      def cmd[Command: ClassTag, Event, State](commandHandler: Command ⇒ PersistentEffect[Event]): ActionHandler[Command, Event, State] = ???
+      def apply[Command: ClassTag, Event, State](commandHandler: ((Command, State, ActorContext[Command]) ⇒ PersistentEffect[Event])): ActionHandler[Command, Event, State] = ???
+      def byState[Command: ClassTag, Event, State](actionHandler: State ⇒ ActionHandler[Command, Event, State]): ActionHandler[Command, Event, State] =
+        new ActionHandler(handler = {
+          case (action, state, ctx) ⇒ actionHandler(state).handler(action, state, ctx)
+        })
```

Comment:
That is an interesting alternative to the `State => Actions`, what are the pros and cons?
Perhaps this is better because it's more explicit in the types: `commandHandler: ActionHandler`.